src/callbacks/perturb_lr.cpp

```diff
@@ -0,0 +1,159 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/comm_impl.hpp"
+#include "lbann/callbacks/perturb_lr.hpp"
+#include "lbann/proto/proto_common.hpp"
+#include "lbann/utils/random_number_generators.hpp"
+#include "lbann/utils/serialize.hpp"
+
+#include <callbacks.pb.h>
+
+#include <algorithm>
+#include <cmath>
+#include <random>
+#include <set>
+#include <sstream>
+#include <utility>
+
+namespace lbann {
+namespace callback {
+
+perturb_lr::perturb_lr(DataType learning_rate_factor,
+                           bool perturb_during_training,
+                           El::Int batch_interval,
+                           std::set<std::string> weights_names)
+  : callback_base(batch_interval),
+    m_learning_rate_factor(learning_rate_factor),
+    m_perturb_during_training(perturb_during_training),
+    m_weights_names(std::move(weights_names)) {}
+
+perturb_lr::perturb_lr()
+  : perturb_lr(0, false, 0)
+{}
+
+template <class Archive>
+void perturb_lr::serialize(Archive & ar) {
+  ar(::cereal::make_nvp(
+       "BaseCallback",
+       ::cereal::base_class<callback_base>(this)),
+     CEREAL_NVP(m_learning_rate_factor),
+     CEREAL_NVP(m_perturb_during_training),
+     CEREAL_NVP(m_weights_names));
+}
+
+void perturb_lr::setup(model* m) {
+  perturb(*m);
+}
+
+void perturb_lr::on_batch_begin(model* m) {
+  const auto& c = m->get_execution_context();
+  if (m_perturb_during_training && c.get_step() > 0) {
```

Comment:
We should only perturb once per interval:
```suggestion
  if (m_perturb_during_training
      && c.get_step() % m_batch_interval == 0
      && c.get_step() > 0) {
```